### PR TITLE
Added type definitions for kerberos npm package

### PR DESCRIPTION
**Pull request by shervinmathieu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

Definition file for kerberos package (https://github.com/mongodb-js/kerberos).

All annotations are pulled from the original source code (https://github.com/mongodb-js/kerberos/blob/master/lib/kerberos.js) with slight changes to conform to DT linting rules.

All functions in Kerberos package either accept a callback function as a parameter or return a Promise if no callback is provided, all functions are declared twice because of this.
Due to Kerberos's implementation, it is possible to call functions with an optional parameter before a required parameter. These functions are declared once more with a different signature to accommodate for this.

### CODE REVIEW

File: types/kerberos/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+ * security could be compromised if you do.
+ *
+ * @param username The Kerberos user name. If no realm is supplied, then the `defaultRealm` will be used.
+ * @param password The password for the user.
+ * @param service The Kerberos service to check access for.
+ * @param defaultRealm The default realm to use if one is not supplied in the user argument.
+ * @param callback
+ * @return returns Promise if no callback passed
+ */
+export function checkPassword(name: string, password: string, service: string, defaultRealm: string): Promise<void>;
```
Comment (armanio123) [inline types/kerberos/index.d.ts]: Based on the documentation `defaultRealm` is optional.

File: types/kerberos/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+export function checkPassword(name: string, password: string, service: string, defaultRealm: string, callback: (err: string) => any): void;
```
Comment (armanio123) [inline types/kerberos/index.d.ts]: `defaultRealm` is optional.

**Review by armanio123 — CHANGES_REQUESTED** [review]

**Comment by velezsarain** [comment]
Hi Armando,
The documentation is wrong. DefaultRealm isn't optional, and passing the wrong argument count or a different argument type than a string will make the function call to be rejected by the built in type checker. Also, the C code bindings that the kerberos package contains expect this value to be a char array, and will error if the value is null (this should never happen because of the type checker).

I don't believe that type definitions should mislead users into using functions with the wrong arguments.

I'm going to fill an issue in the kerberos repository to check if this was intentional or is an oversight.

**Comment by velezsarain** [comment]
I've filled an issue in the kerberos package (https://github.com/mongodb-js/kerberos/issues/91)

I will wait for the package maintainer to respond to check if the type definition should be changed or not.

**Pull request closed without merging**